infra/events/src/main/java/com/simprints/infra/events/session/SessionEventRepositoryImpl.kt

```diff
@@ -20,98 +20,114 @@ internal class SessionEventRepositoryImpl @Inject constructor(
         val session = sessionDataCache.eventScope ?: localSessionScope()
-        return session != null
+        session != null
     }
 
-    override suspend fun getCurrentSessionScope(): EventScope {
-        val cachedScope = sessionDataCache.eventScope
-        if (cachedScope != null) return cachedScope
-
-        val restoredScope = localSessionScope()
-        if (restoredScope != null) {
-            Simber.w("Restored session from DB")
-            return restoredScope
-        }
-
-        Simber.w("Creating new session DB")
-        return createSession()
+    override suspend fun getCurrentSessionScope(): EventScope = withLockedContext {
+        getCurrentScopeInternal()
     }
 
-    override suspend fun removeLocationDataFromCurrentSession() {
-        val sessionScope = getCurrentSessionScope()
+    override suspend fun removeLocationDataFromCurrentSession() = withLockedContext {
+        val sessionScope = getCurrentScopeInternal()
         if (sessionScope.payload.location != null) {
             val updatedSessionScope = sessionScope.copy(
                 payload = sessionScope.payload.copy(location = null)
             )
-            saveSessionScope(updatedSessionScope)
+            saveSessionScopeInternal(updatedSessionScope)
         }
     }
 
-    override suspend fun getEventsInCurrentSession(): List<Event> {
-        val sessionId = getCurrentSessionScope().id
-        return eventsLock.withLock {
-            if (sessionDataCache.eventCache.isEmpty()) {
-                loadEventsIntoCache(sessionId)
-            }
-            sessionDataCache.eventCache.values.toList()
+    override suspend fun getEventsInCurrentSession(): List<Event> = withLockedContext {
+        val sessionId = getCurrentScopeInternal().id
+
+        if (sessionDataCache.eventCache.isEmpty()) {
+            loadEventsIntoCache(sessionId)
         }
+        sessionDataCache.eventCache.values.toList()
     }
 
-    override suspend fun addOrUpdateEvent(event: Event) = eventsLock.withLock {
+    override suspend fun addOrUpdateEvent(event: Event) = withLockedContext {
         val currentEvents = sessionDataCache.eventCache.values.toList()
         val savedEvent = eventRepository.addOrUpdateEvent(
-            getCurrentSessionScope(),
+            getCurrentScopeInternal(),
             event,
-            currentEvents
+            currentEvents,
         )
         sessionDataCache.eventCache[savedEvent.id] = savedEvent
     }
 
-    private suspend fun localSessionScope() = eventRepository
+    override suspend fun closeCurrentSession(reason: EventScopeEndCause?) = withLockedContext {
+        eventRepository.closeEventScope(getCurrentScopeInternal(), reason)
+        sessionDataCache.eventCache.clear()
+        sessionDataCache.eventScope = null
+    }
+
+    /**
+     * Scope for internal extension functions that must be called within locked context.
+     */
+    private object LockedContext
```

Comment:
Interesting approach! So, if I got this right we have the internal functions explicitly defined as extensions of this context and additionally all blocks executed with withLockedContext() are also executed in this context (even through not explicitly defined as extensions since they are lambdas)?

Reply:
Yes, internal functions are implemented as extensions to force the caller to think if the called function uses the parts that are locked with mutex. The "context" in this case does not do anything in particular besides being the receiver of the internal functions.